File: src/packages/control-ui/src/api/desktop_client_manager.ts
export const SystemProxyMode = {
  Direct: 'direct',
  Pac: 'pac',
  Global: 'global',
  Manual: 'manual',
} as const;
export type SystemProxyMode = (typeof SystemProxyMode)[keyof typeof SystemProxyMode];

export interface DesktopStatus {
  mode: SystemProxyMode;
  activeProfile: string;
  httpPort: number;
  socks5Port: number;
  isConnected: boolean;
  pacUrl?: string | undefined;
}

export class DesktopClientManager {
  private mode: SystemProxyMode = SystemProxyMode.Direct;
  private activeProfile: string = 'default';
  public httpPort: number;
  public socks5Port: number;
  private isConnected: boolean = false;
  private pacUrl?: string;

  constructor(httpPort: number, socks5Port: number) {
    this.httpPort = httpPort;
    this.socks5Port = socks5Port;
  }

  setProxyMode(mode: SystemProxyMode, pacUrl?: string): void {
    if (mode === SystemProxyMode.Pac && !pacUrl && !this.pacUrl) {
      throw new Error('PAC mode requires a valid PAC URL');
    }

    this.mode = mode;
    if (pacUrl) {
      this.pacUrl = pacUrl;
    }
  }

  switchProfile(profileId: string): void {
    this.activeProfile = profileId;
  }

  setConnected(connected: boolean): void {
    this.isConnected = connected;
  }

  getStatus(): DesktopStatus {
    return {
      mode: this.mode,
      activeProfile: this.activeProfile,
      httpPort: this.httpPort,
      socks5Port: this.socks5Port,
      isConnected: this.isConnected,
      pacUrl: this.pacUrl,
    };
  }

  handleIpcCommand(
    command: string,
    payload: Record<string, unknown> = {},
  ): DesktopStatus | IpcCommandResult {
    switch (command) {
      case 'get_status':
        return this.getStatus();
      case 'set_mode': {
        const modeStr = typeof payload.mode === 'string' ? payload.mode : undefined;
        if (!modeStr) throw new Error('Missing mode');
        let mode: SystemProxyMode;
        switch (modeStr.toLowerCase()) {
          case 'direct':
            mode = SystemProxyMode.Direct;
            break;
          case 'pac':
            mode = SystemProxyMode.Pac;
            break;
          case 'global':
            mode = SystemProxyMode.Global;
            break;
          case 'manual':
            mode = SystemProxyMode.Manual;
            break;
          default:
            throw new Error('Invalid proxy mode');
        }
        const pacUrl =
          typeof payload.pac_url === 'string'
            ? payload.pac_url
            : typeof payload.pacUrl === 'string'
              ? payload.pacUrl
              : undefined;
        this.setProxyMode(mode, pacUrl);
        return { success: true };
      }
      case 'switch_profile': {
        const profile = typeof payload.profile === 'string' ? payload.profile : undefined;
        if (!profile) throw new Error('Missing profile');
        this.switchProfile(profile);
        return { success: true, profile };
      }
      default:
        throw new Error(`Unknown command: ${command}`);
    }
  }
}

/** Result envelope for commands handled through the desktop IPC bridge. */
export interface IpcCommandResult {
  success: boolean;
  profile?: string;
}

File: labs/mobile/donor-archive/tunnel/DesktopClientManager.kt
package com.luminet.android.tunnel

enum class SystemProxyMode {
    DIRECT,
    PAC,
    GLOBAL,
    MANUAL
}

data class DesktopStatus(
    val mode: SystemProxyMode,
    val activeProfile: String,
    val httpPort: Int,
    val socks5Port: Int,
    val isConnected: Boolean,
    val pacUrl: String?
)

class DesktopClientManager(val httpPort: Int, val socks5Port: Int) {
    var mode: SystemProxyMode = SystemProxyMode.DIRECT
        private set
    var activeProfile: String = "default"
        private set
    var isConnected: Boolean = false
        private set
    var pacUrl: String? = null
        private set

    fun setProxyMode(newMode: SystemProxyMode, url: String? = null) {
        if (newMode == SystemProxyMode.PAC && url == null && pacUrl == null) {
            throw IllegalArgumentException("PAC mode requires PAC URL")
        }
        mode = newMode
        if (url != null) pacUrl = url
    }

    fun switchProfile(profile: String) {
        activeProfile = profile
    }

    fun setConnected(connected: Boolean) {
        isConnected = connected
    }

    fun getStatus(): DesktopStatus {
        return DesktopStatus(
            mode = mode,
            activeProfile = activeProfile,
            httpPort = httpPort,
            socks5Port = socks5Port,
            isConnected = isConnected,
            pacUrl = pacUrl
        )
    }

    fun handleIpcCommand(command: String, payload: Map<String, Any>): Map<String, Any> {
        return when (command) {
            "get_status" -> mapOf(
                "mode" to mode.name,
                "active_profile" to activeProfile,
                "is_connected" to isConnected
            )
            "set_mode" -> {
                val mStr = payload["mode"] as? String ?: throw IllegalArgumentException("Missing mode")
                val url = payload["pac_url"] as? String
                setProxyMode(SystemProxyMode.valueOf(mStr.uppercase()), url)
                mapOf("success" to true)
            }
            "switch_profile" -> {
                val prof = payload["profile"] as? String ?: throw IllegalArgumentException("Missing profile")
                switchProfile(prof)
                mapOf("success" to true, "profile" to prof)
            }
            else -> throw IllegalArgumentException("Unknown IPC command: $command")
        }
    }
}
